src/server/api/v1/utils.py

```diff
@@ -0,0 +1,102 @@
+            'iat': datetime.datetime.utcnow(),
+            'sub': user_id
+        }
+        return jwt.encode(
+            payload,
+            SECRET_KEY,
+            algorithm='HS256')
+    except Exception as e:
+        return e
+
+
+def decode_auth_token(auth_token):
+    try:
+        payload = jwt.decode(auth_token, SECRET_KEY)
+        return payload['sub']
+    except jwt.ExpiredSignatureError:
+        return "Signature expired. Please sign-in again"
+    except jwt.InvalidTokenError:
+        return "Invalid token. Please sign-in again"
+
+
+def default(req):
+    req.serv_reply(200, {'message': 'Welcome to Warship Battle! Visit \'/register\' to get registered!'})
+
+
+def path_not_found(req):
+    req.serv_reply(404, {'message': 'Requested path is not found, please check the endpoint and try again'})
+
+
+def login(req, data):
+    user = db_helper.db_connection.query(UserModel).filter_by(nickname=data['nickname']).first()
+    if not (db_helper.UserDBHelper.user_exists(data) and check_password_hash(user.passwd, data['password'])):
+        req.serv_reply(401, {'message': 'Unauthorized. Username or password is invalid'})
+    else:
+        token = encode_auth_token(user.id)
+
+        # data to be updated in the DB {<field_name>: <value>}
+        data = {'token': token}
+        db_helper.UserDBHelper.update(user.id, data)
+        db_helper.db_connection.commit()
+        req.serv_reply(200, {'token': token.decode('utf-8')})
+
+
+def register(req, data):
+    if not db_helper.UserDBHelper.user_exists(data):
+        user_obj = db_helper.UserDBHelper(db_helper.db_connection)
+        passwd = generate_password_hash(data['password'])
+        user_obj.create(data['nickname'], passwd)
+        req.serv_reply(201, {'message': 'User successfully created'})
+    else:
+        req.serv_reply(409, {'message': 'Username is already taken'})
+
+
+def users(req):
+    users = db_helper.db_connection.query(UserModel).all()
+    response = {'users': {}}
+    for i in range(len(users)):
+        data = {'id': users[i].id, 'name': users[i].nickname}
+        response['users'][i] = data
+    req.serv_reply(200, response)
+
+
+def get_user(req, id):
+    try:
+        user = db_helper.db_connection.query(UserModel).filter_by(id=id).first()
+        response = {'id': user.id, 'name': user.nickname}
+        req.serv_reply(200, response)
+    except AttributeError:
+        req.serv_reply(400, {'message': f'User with id {id} is not found'})
+
+
+def dispatch_handler(req_obj, *args, **kwargs):
+    # I believe this is not the best solution to handle user id's in such way
+    url_parts = req_obj.path.split('/')
+    if req_obj.command == 'GET' and url_parts[-1].isdigit():
+        return get_user(req_obj, (url_parts[-1]))
+
+    elif req_obj.command == 'GET':
+        return {
+            '/': default,
+            '/users': users,
+            }.get(req_obj.path, path_not_found)(req_obj, *args, **kwargs)
```

Comment:
We will catch the error if fro example path won`t be found but *args or **kwargs will be present because in such case we will return 'path_not_found(req_obj, *args, **kwargs)' but the interface of the function is 'def path_not_found(req):'.
The same for other functions in case if *args will be passed.

Reply:
Done